### PR TITLE
Improve cache handling on the frontend, cache executions on the backend, and improve controls on the exec pane

### PR DESCRIPTION
**Pull request by jeremy-rifkin**

### CODE REVIEW

**Comment by jeremy-rifkin** [comment]
Should be good. I'll test thoroughly tomorrow and fix test cases / merge conflicts and whatnot.

**Comment by jeremy-rifkin** [comment]
This should close #3030

File: lib/base-compiler.ts [hunk 1]
```diff
@@ -1695,11 +1696,13 @@ export class BaseCompiler implements ICompiler {
         };
     }
 
-    async getOrBuildExecutable(key) {
+    async getOrBuildExecutable(key, bypassCache?: BypassCacheControl) {
         const dirPath = await this.newTempDir();
 
-        const buildResults = await this.loadPackageWithExecutable(key, dirPath);
-        if (buildResults) return buildResults;
+        if (!bypassCache || !(bypassCache & BypassCacheControl.Compilation)) {
```
Comment (partouf) [inline lib/base-compiler.ts:1702]: can you write this with && ? using a bitwise operator for the first time in the codebase seems confusing

Comment (jeremy-rifkin) [reply lib/base-compiler.ts:1702]: `bypassCache` is a flag enum value, this check could be abstracted to a function but I don't think it'd be entirely warranted to do so.

Comment (jeremy-rifkin) [reply lib/base-compiler.ts:1702]: Well, thinking about it more I can't imagine it ever making sense to recompile but not also re-execute, so maybe everything can be rephrased not in terms of a flag enum.

**Comment by jeremy-rifkin** [comment]
Alright, tested locally and everything is behaving how I want it. Does in fact close 3030.

**Comment by jeremy-rifkin** [comment]
The interface change in this PR gets rid of the circle arrow button in the top-right corner and adds a re-run button next to the clear cache and recompile button
image

File: lib/handlers/compile.ts [hunk 2]
```diff
@@ -81,20 +82,15 @@ function initialise(compilerEnv: CompilationEnvironment) {
...
     source: string;
     options: string[];
     backendOptions: Record<string, any>;
     filters: ParseFiltersAndOutputOptions;
-    bypassCache: boolean;
+    bypassCache: BypassCache;
     tools: any;
     executionParameters: ExecutionParams;
-    libraries: any[];
+    libraries: CompileChildLibraries[];
```
Comment (RubenRBS) [inline lib/handlers/compile.ts:93]: ❤️

File: docs/API.md [hunk 3]
```diff
@@ -121,7 +121,17 @@ The filters are a JSON object with `true`/`false` values. If not supplied, defau
 filters override their default values. The `compilerOptions` is used to pass extra arguments to the back end, and is
 probably not useful for most REST users.
 
-To force a cache bypass, set `bypassCache` in the root of the request to `true`.
+To force a cache bypass, `bypassCache` can be set. This accepts an enum value according to:
```
Comment (RubenRBS) [inline docs/API.md:124]: What happens to thirdparty users setting this to true? Does the new api allow for that fallback?

Comment (RubenRBS) [reply docs/API.md:124]: Oh just seen https://github.com/compiler-explorer/compiler-explorer/pull/5111/files#diff-e16897a5a4f4bbb218c2be258bb4908e073a4d3272181618dbe77c1cf559a234R92, I guess that's what backwards compat means here :)

File: views/templates/panes/executor.pug [hunk 4]
```diff
@@ -58,7 +56,9 @@
     .compiler-output
     .execution-output
   .bottom-bar.bg-light
-    button.btn.btn-sm.btn-light.trigger-compilation(title="Trigger compilation")
+    button.btn.btn-sm.btn-light.rerun(title="Rerun")
+      span.fas.fa-circle-play
```
Comment (RubenRBS) [inline views/templates/panes/executor.pug:60]: I'm ok with this one, but fa-replay could also be an option. Whatever you like :)

**Review by RubenRBS — APPROVED** [review]
Great work!

File: docs/API.md (hunk 3 above)
Comment (jeremy-rifkin) [reply docs/API.md:124]: Yes! I went ahead and added some comments here and in the compiler request handler code.

**Pull request merged**